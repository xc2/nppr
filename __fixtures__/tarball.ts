export const BasicTarballPath = new URL("./barhop-0.0.0-PLACEHOLDER.tgz", import.meta.url).pathname;
