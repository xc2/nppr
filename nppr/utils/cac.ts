import { EOL } from "node:os";
import * as process from "node:process";
import type { CAC, Command } from "cac";
import stringWidth from "string-width";

export interface CliCommand {
  (get: CAC["command"]): Command;
}

export const registerCommand = (cli: CAC, name: string, cmd: CliCommand) => {
  const register: CAC["command"] = (rawName, ...args) => {
    return cli.command([name, rawName].filter(Boolean).join(" "), ...args);
  };
  return cmd(register);
};

export class ArgumentsError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ArgumentsError";
  }
}

export function getOptionsForUsage(cli: CAC) {
  const cmd = cli.matchedCommand || cli.globalCommand;
  const { options: globalOptions } = cli.globalCommand;
  let options = cmd.isGlobalCommand ? globalOptions : [...cmd.options, ...(globalOptions || [])];

  if (!cmd.isGlobalCommand && !cmd.isDefaultCommand) {
    options = options.filter((option) => option.name !== "version");
  }

  const maxColumn = getMaxLength(options.map((option) => option.rawName));
  return {
    maxColumn,
    options,
  };
}

function getMaxLength(arr: string[]) {
  return arr.reduce((a, b) => Math.max(a, b.length), 0);
}

export function indent(str: string, count = 2) {
  const spaces = " ".repeat(count);
  return str.replace(/^(?=.+$)/gm, spaces);
}

export function trimLeft(str: string) {
  return str.replace(/^\s+/g, "");
}

export function simplyFold(str: string, ind = 0, columns?: number) {
  const renderColumns = (columns || process.stdout.columns || 80) - ind;
  const lines = str.split(/(\r\n|\r|\n)/g);
  const spaces = " ".repeat(ind);
  const fin: string[] = [];
  for (let i = 0; i < lines.length; i += 2) {
    const line = lines[i];
    const br = lines[i + 1] ?? "";
    const words = line.split(/(\s+)/g);
    let currentLine = "";
    let currentLength = -1;
    for (let j = 0; j < words.length; j += 2) {
      const word = words[j] + (words[j + 1] ?? "");
      const len = stringWidth(word);
      const nextLen = currentLength + len;
      if (nextLen > renderColumns) {
        fin.push(spaces + currentLine + EOL);
        currentLine = word;
        currentLength = len;
      } else {
        currentLine += word;
        currentLength = nextLen;
      }
    }
    if (currentLine) {
      fin.push(spaces + currentLine);
    }
    fin.push(br);
  }

  return fin.join("");
}
