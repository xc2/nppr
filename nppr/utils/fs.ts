import { createReadStream, createWriteStream } from "node:fs";
import { tryToNumber } from "./lang";

export function outputStream(filepath: string | number) {
  filepath = tryToNumber(filepath);

  if (typeof filepath === "string") {
    // file path
    return createWriteStream(filepath);
  }
  if (typeof filepath === "number") {
    const builtins: Record<string, NodeJS.WritableStream> = {
      [process.stdout.fd]: process.stdout,
      [process.stderr.fd]: process.stderr,
    };
    // file descriptor
    return builtins[filepath] ?? createReadStream("", { fd: filepath });
  }
  throw new Error(`Invalid output filepath: ${filepath}`);
}
