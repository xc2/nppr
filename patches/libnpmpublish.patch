diff --git a/lib/provenance.js b/lib/provenance.js
index cead8da6be34fe3940ef7552c2b4a955efec1b1f..516dabb1727f92cde27ece2d1415d5cbaafc39d6 100644
--- a/lib/provenance.js
+++ b/lib/provenance.js
@@ -198,32 +198,34 @@ const generateProvenance = async (subject, opts) => {
 
 const verifyProvenance = async (subject, provenancePath) => {
   let provenanceBundle
-  try {
-    provenanceBundle = JSON.parse(await readFile(provenancePath))
-  } catch (err) {
-    err.message = `Invalid provenance provided: ${err.message}`
-    throw err
+  if (typeof provenancePath === 'string') {
+    try {
+      provenanceBundle = JSON.parse(await readFile(provenancePath))
+    } catch (err) {
+      err.message = `Invalid provenance provided: ${err.message}`
+      throw err
+    }
+  } else if (provenancePath && typeof provenancePath === 'object') {
+    provenanceBundle = provenancePath
   }
 
   const payload = extractProvenance(provenanceBundle)
   if (!payload.subject || !payload.subject.length) {
     throw new Error('No subject found in sigstore bundle payload')
   }
-  if (payload.subject.length > 1) {
-    throw new Error('Found more than one subject in the sigstore bundle payload')
-  }
 
-  const bundleSubject = payload.subject[0]
-  if (subject.name !== bundleSubject.name) {
-    throw new Error(
-      `Provenance subject ${bundleSubject.name} does not match the package: ${subject.name}`
-    )
+  const bundleSubject = payload.subject.find((s) => s.name === subject.name)
+  if (!bundleSubject) {
+    throw new Error(`No subject found in sigstore bundle payload matching the package: ${subject.name}`)
   }
   if (subject.digest.sha512 !== bundleSubject.digest.sha512) {
     throw new Error('Provenance subject digest does not match the package')
   }
 
-  await sigstore.verify(provenanceBundle)
+  if (process.env.SKIP_VERIFY_SIGSTORE_USE_FOR_TESTING_ONLY_OR_IF_YOU_KNOW_WHAT_YOU_ARE_DOING !== 'true') {
+    await sigstore.verify(provenanceBundle)
+  }
+
   return provenanceBundle
 }
 
