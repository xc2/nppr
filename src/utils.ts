export function dirtyTimestampToDate(ts: string, ratio = 1) {
  if (!ts) {
    return null;
  }
  const d = new Date(Number(ts) * ratio);
  if (d.getTime()) {
    return d;
  }
  return null;
}

export function toYMDHMS(date: Date) {
  return date
    .toISOString()
    .replace(/\..+$/, "")
    .replace(/[^0-9]/g, "");
}

export async function destructPromise<Data, Err extends Error = Error>(
  promise: PromiseLike<Data>
): Promise<[false, Err] | [true, Data]> {
  try {
    const data = await promise;
    return [true, data];
  } catch (e) {
    return [false, e as Err];
  }
}
