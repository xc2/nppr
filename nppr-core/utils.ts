import { createReadStream } from "node:fs";
import npa from "npm-package-arg";
import coerceVersion from "semver/functions/coerce";
import { tryToNumber } from "./internals/lang";
import { template as _template } from "./internals/lang/template";
import type { Manifest } from "./internals/package";
import { bufferToReadable, toReadableStream } from "./internals/stream";

export type InputSource =
  | string
  | number
  | Blob
  | File
  | ReadableStream
  | NodeJS.ReadableStream
  | ArrayBufferLike
  | ArrayBufferView;

export function inputSource(source: InputSource): ReadableStream {
  if (typeof source === "string") {
    source = tryToNumber(source);
    if (typeof source === "string") {
      // file path
      return toReadableStream(createReadStream(source));
    }
  }
  if (typeof source === "number") {
    const builtins: Record<string, NodeJS.ReadableStream> = {
      [process.stdin.fd]: process.stdin,
    };
    // file descriptor
    return toReadableStream(builtins[source] ?? createReadStream("", { fd: source }));
  }
  if ("pipe" in source) {
    // nodejs stream
    return toReadableStream(source);
  }
  if ("arrayBuffer" in source) {
    if (typeof source.stream === "undefined") {
      throw new Error("source might be a Blob or File, but it does not have a stream method");
    }
    return source.stream();
  }
  if ("getReader" in source) {
    // ReadableStream
    return source;
  }

  // ArrayBufferView | ArrayBufferLike
  return bufferToReadable(source);
}

export const inlineTemplate = _template.bind(null, /\[([^\]]+)]/g);
export const template = _template.bind(null, /\{\{([^}]+)}}/g);

export function toPurl(manifest: Pick<Manifest, "name" | "version">) {
  const spec = npa.resolve(manifest.name, manifest.version);
  // @ts-ignore
  return npa.toPurl(spec) as string;
}

export function getPackageTokens(manifest: Manifest) {
  const { name, version } = manifest;
  return {
    ...manifest,
    names: extractName(name),
    versions: extractVersion(version),
  };
}

export function extractName(name: string) {
  if (name.startsWith("@")) {
    const [scope, unscoped] = name.slice(1).split("/");
    return {
      scope,
      unscoped,
      forPath: `${scope}-${unscoped}`,
      forUnscoped: `${scope}__${unscoped}`,
    };
  }
  return { scope: "", unscoped: name, "for-path": name, "for-unscoped": name };
}
export function extractVersion(version: string) {
  const v = coerceVersion(version, { includePrerelease: true, loose: true });
  if (!v) {
    return {};
  }
  return {
    major: v.major,
    minor: v.minor,
    patch: v.patch,
    version: `${v.major}.${v.minor}.${v.patch}`,
    prerelease: v.prerelease.join("."),
    prereleases: v.prerelease,
    build: v.build.join("."),
    builds: v.build,
  };
}

export function getPublishManifestFields() {
  return [
    "name",
    "version",
    "bin",
    "description",
    "keywords",
    "homepage",
    "bugs",
    "license",
    "author",
    "contributors",
    "funding",
    "engines",
    "repository",
    "dependencies",
    "peerDependencies",
    "peerDependenciesMeta",
    "optionalDependencies",
    "os",
    "cpu",
    // yarnpkg-specific fields below
    "libc",
    "languageName",
    "dependenciesMeta",
    "preferUnplugged",
  ];
}

export const NPPR_USER_AGENT = "nppr-core/v1 (+https://github.com/xc2/nppr)";
