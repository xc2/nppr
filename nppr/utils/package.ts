import { createWriteStream } from "node:fs";
import * as NodePath from "node:path";
import { Writable } from "node:stream";
import { type Manifest, getManifest, inlineTemplate, inputSource, tryToNumber } from "nppr-core";

export function pathInfo(path: string, cwd: string) {
  const abs = NodePath.resolve(cwd, path);
  const rel = NodePath.relative(cwd, abs);
  let ext = NodePath.extname(abs);
  let base = NodePath.basename(abs, ext);
  if (NodePath.extname(base) === ".tar") {
    ext = ".tar" + ext;
    base = NodePath.basename(base, ext);
  }
  return { fullpath: abs, path: NodePath.dirname(rel), extname: ext, basename: base };
}

export function packageName(name: string) {
  if (name.startsWith("@")) {
    const [scope, unscoped] = name.slice(1).split("/");
    return { scope, unscoped, pathPart: `${scope}-${unscoped}` };
  }
  return { scope: "", unscoped: name, pathPart: name };
}

export class Package {
  public readonly input: string | number;
  public readonly pathInfo?: ReturnType<typeof pathInfo>;
  protected _source?: ReadableStream;
  protected _manifest?: Promise<Manifest>;

  constructor(
    input: string | number,
    public readonly cwd = process.cwd()
  ) {
    const p = tryToNumber(input);
    if (typeof p === "string") {
      this.pathInfo = pathInfo(p, this.cwd);
      this.input = this.pathInfo.fullpath;
    } else {
      this.input = p;
    }
  }

  tee() {
    const [a, b] = this.source.tee();
    this._source = a;
    return b;
  }

  get source() {
    this._source = this._source ?? inputSource(this.input);
    return this._source;
  }

  set source(source: ReadableStream) {
    if (this._source !== source) {
      this._source = source;
      this._manifest = undefined;
    }
  }

  get manifest() {
    try {
      this._manifest = this._manifest || getManifest(this.tee());
      return this._manifest;
    } catch (e) {
      return Promise.reject(e);
    }
  }

  async output(p?: string) {
    const outputPath = await this.getOutputPath(p);
    return this.tee().pipeTo(Writable.toWeb(createWriteStream(outputPath)));
  }

  // TODO: move along and add tests
  async getOutputPath(p: string = "[path]/[name]-[version][extname]") {
    const manifest = await this.manifest;
    const name = packageName(manifest.name);
    const variables = {
      ...manifest,
      scope: name.scope,
      unscoped: name.unscoped,
      fullpath: this.pathInfo?.fullpath ?? "",
      path: this.pathInfo?.path ?? "",
      extname: this.pathInfo?.extname ?? ".tgz",
      basename: this.pathInfo?.basename ?? `${name.pathPart}-${manifest.version}`,
    };
    const _escape = (value: any, key: string) => {
      if (key === "name") {
        return packageName(value).pathPart;
      }
      return `${value}`.replace(/[^0-9a-zA-Z-._]/g, "_");
    };
    return NodePath.resolve(this.cwd, inlineTemplate(p, { escape: _escape })(variables));
  }
}
