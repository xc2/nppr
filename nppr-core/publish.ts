import { readFile } from "node:fs/promises";
import { type PublishOptions as _PublishOptions, publish as npmpublish } from "libnpmpublish";
import { type Manifest, getManifest } from "./internals/package";
import { readableToBuffer } from "./internals/stream";
import { type InputSource, NPPR_USER_AGENT, getPublishManifestFields, inputSource } from "./utils";

export interface ManifestPublishOptions {
  keepFields?: true | string[];
  additionalFields?: Partial<Manifest> | ((manifest: Manifest) => Partial<Manifest>);
}

export interface PublishOptions
  extends Omit<_PublishOptions, "provenanceFile" | "provenance" | "defaultTag"> {
  manifest?: ManifestPublishOptions;
  provenanceBundle?: string | object;
  tag?: string;
  token?: string;
}

export async function publish(
  tarball: InputSource,
  { manifest: manifestOptions, provenanceBundle, ...publishOptions } = {} as PublishOptions
) {
  const [source1, source2] = inputSource(tarball).tee();
  const packageJson = await getManifest(source1);
  const manifest = getPublishManifest(packageJson, manifestOptions);
  if (provenanceBundle) {
    const repo = await getRepoFromProvenance(provenanceBundle);
    if (repo) {
      addRepoInfoToManifest(manifest, repo);
    }
  }
  const config = getPublishConfig({ ...packageJson, ...manifest }, publishOptions);
  // @ts-ignore
  config.provenanceFile = provenanceBundle;

  return npmpublish(
    manifest as any,
    // TODO: libnpmpublish only accepts Buffer
    Buffer.from(await readableToBuffer(source2)),
    config
  );
}

export function getPublishManifest(manifest: Manifest, options = {} as ManifestPublishOptions) {
  manifest = JSON.parse(JSON.stringify(manifest));
  const r = {} as Manifest;
  if (options.keepFields === true) {
    Object.assign(r, manifest);
  } else {
    const extraFields = Array.isArray(options.keepFields) ? options.keepFields : [];
    for (const field of new Set([...(extraFields || []), ...getPublishManifestFields()])) {
      if (field in manifest) {
        r[field] = manifest[field];
      }
    }
  }
  const additionalFields =
    typeof options.additionalFields === "function"
      ? options.additionalFields({ ...manifest })
      : options.additionalFields;
  return { ...r, ...additionalFields };
}

export function getPublishConfig(
  manifest: Manifest,
  { tag, ...options } = {} as Omit<PublishOptions, "manifest" | "provenanceBundle">
): _PublishOptions {
  return {
    ...manifest.publishConfig,
    npmVersion: NPPR_USER_AGENT,
    ...options,
    defaultTag: tag || "latest",
    provenance: false,
    forceAuth: {
      token: options._authToken || options.token,
    },
  };
}

export async function getRepoFromProvenance(provenance: string | object) {
  if (typeof provenance === "string") {
    provenance = JSON.parse(await readFile(provenance, "utf8"));
  }
  const p = provenance as any;
  const payloadSource = JSON.parse(Buffer.from(p.dsseEnvelope.payload, "base64").toString("utf-8"));
  if (payloadSource?.predicate?.buildDefinition?.resolvedDependencies?.[0]) {
    const dep = payloadSource.predicate.buildDefinition.resolvedDependencies[0];
    return getRepositoryInfo(dep.uri, dep.digest.gitCommit);
  } else if (payloadSource?.predicate?.materials?.[0]) {
    const dep = payloadSource.predicate.materials[0];
    return getRepositoryInfo(dep.uri, dep.digest.sha1);
  }
  return null;
}

interface Repo {
  repository: string;
  commit: string;
  homepage?: string;
  issues?: string;
}

function getRepositoryInfo(repo: string, commit: string): Repo | null {
  if (!repo || !commit) return null;
  const u = new URL(repo);
  if (u.pathname.includes("@refs")) {
    const [name, ref] = u.pathname.split("@refs/");
    u.pathname = name;
  }
  const fin = {
    repository: u.toString(),
    commit: commit,
  } as Repo;
  if (u.host === "github.com") {
    const webUrl = `https://github.com${u.pathname}`;
    fin.repository = `git+https://github.com${u.pathname}.git`;
    fin.homepage = webUrl;
    fin.issues = `${webUrl}/issues`;
  }
  return fin;
}
function addRepoInfoToManifest(manifest: Manifest, repo: Repo) {
  if (!manifest.repository) {
    manifest.repository = { type: "git", url: repo.repository };
  } else if (typeof manifest.repository !== "string") {
    if (!manifest.repository.url) {
      manifest.repository.url = repo.repository;
      manifest.repository.type = "git";
    }
  }
  if (!manifest.homepage) {
    manifest.homepage = repo.homepage;
  }
  if (!manifest.bugs) {
    manifest.bugs = repo.issues;
  } else if (typeof manifest.bugs !== "string") {
    if (!manifest.bugs.url) {
      manifest.bugs.url = repo.issues;
    }
  }
  if (!manifest.gitHead) {
    manifest.gitHead = repo.commit;
  }
}
