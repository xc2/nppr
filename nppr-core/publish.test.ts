import { publish as _publish } from "libnpmpublish";
import { mockImplementation, test } from "tests/vitest";
import { afterAll, describe, expect, vi } from "vitest";
import { BasicTarballPath } from "../tests/__fixtures__/tarball";
import type { Manifest } from "./internals/package";
import { getPublishConfig, getPublishManifest, publish } from "./publish";
import { NPPR_USER_AGENT } from "./utils";

vi.mock("libnpmpublish", { spy: true });

afterAll(() => {
  vi.doUnmock("libnpmpublish");
});
describe("publish", () => {
  test("should publish a package", async ({ signal }) => {
    mockImplementation(_publish as any, (...args: any[]) => args, signal);

    const r = await publish(BasicTarballPath, {
      token: "foo",
      tag: "bar",
    });
    expect(r).toHaveLength(3);
    // @ts-ignore
    const [manifest, tarball, config] = r;
    expect(manifest).toMatchSnapshot();
    expect(tarball).toBeInstanceOf(Buffer);
    expect(tarball).toHaveLength(15792);
    expect(config).toMatchObject({ forceAuth: { token: "foo" }, defaultTag: "bar" });
  });
});

describe("getPublishManifest", () => {
  const defaultAllows = { name: "foo", version: "1.0.0" } as Manifest;
  const manifest = { ...defaultAllows, foo: "bar", hello: "world" } as Manifest;
  test("should keep all fields when `keepFields` is true", () => {
    expect(getPublishManifest(manifest, { keepFields: true })).toEqual(manifest);
  });
  test("should remove unwanted fields", () => {
    expect(getPublishManifest(manifest)).toEqual(defaultAllows);
  });
  test("should allow keep specific fields", () => {
    expect(getPublishManifest(manifest, { keepFields: ["foo"] })).toEqual({
      ...defaultAllows,
      foo: "bar",
    });
  });
  test("should allow additional fields as object", () => {
    const add = { a: 1 };
    expect(getPublishManifest(manifest, { additionalFields: add })).toEqual({
      ...defaultAllows,
      ...add,
    });
  });
  test("should allow additional fields as function", () => {
    expect(
      getPublishManifest(manifest, {
        additionalFields: (raw) => {
          return { a: raw.name };
        },
      })
    ).toEqual({
      ...defaultAllows,
      a: defaultAllows.name,
    });
  });
});

describe("getPublishConfig", () => {
  const manifest = { name: "foo", version: "1.0.0" } as Manifest;
  test("should have npmVersion set", () => {
    expect(getPublishConfig(manifest)).toMatchObject({
      npmVersion: NPPR_USER_AGENT,
    });
  });
});
