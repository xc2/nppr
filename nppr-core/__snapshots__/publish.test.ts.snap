// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`publish > should publish a package 1`] = `
{
  "name": "barhop",
  "version": "0.0.0-PLACEHOLDER",
}
`;
